File: week_1/percolation.rb
require 'byebug'
require_relative 'weighted_quick_union'

class Percolation
  attr_accessor :size, :strategy, :sites

  def initialize(n)
    @row_length = n
    @size = n * n
    @strategy = WeightedQuickUnion.new(@size + 2)
    @sites = Array.new(@row_length) { Array.new(@row_length, 0) }
    @virtual_top = @size
    @virtual_bottom = @size + 1

    @sites.each_with_index do |ary, index|
      @strategy.union(@virtual_top, index)
      @strategy.union(@virtual_bottom, @size - 1 - index)
    end
  end

  def is_open?(row, col)
    @sites[row - 1][col - 1] == 1
  end

  def open(row, col)
    row_index = row - 1
    col_index = col - 1

    unless is_open?(row, col)
      site_id = (row_index * @row_length) + col_index

      neighbors = [[row_index-1, col_index],
                   [row_index+1, col_index],
                   [row_index, col_index-1],
                   [row_index, col_index+1]]

      neighbors.each do |neighbor|
        next if neighbor[0] < 0 || neighbor[0] > @row_length || neighbor[1] < 0 || neighbor[1] > @row_length

        if is_open?(neighbor[0], neighbor[1])
          # why did I need to normalize here?
          neighbor_site_id = ((neighbor[0] - 1) * @row_length) + (neighbor[1] - 1)

          if !@strategy.connected?(site_id, neighbor_site_id)
            @strategy.union(site_id, neighbor_site_id)
          end

          if is_full?(neighbor[0] + 1, neighbor[1] + 1)
            @strategy.union(@virtual_top, site_id)
          end
        end
      end

      @sites[row_index][col_index] = 1
    end
  end

  def is_full?(row, col)
    row_index = row - 1
    col_index = col - 1
    site_id = (row_index * @row_length) + col_index

    is_open?(row, col) && @strategy.connected?(site_id, @virtual_top)
  end

  def percolates?
    @strategy.connected?(@virtual_top, @virtual_bottom)
  end

  def monte_carlo
    @sites.shuffle.each do |site|
      open(site[0], site[1])

      break if percolates?
    end
  end
end

File: submissions/Percolation.java
public class Percolation {
    private static final int BLOCKED = 0;
    private static final int OPEN = 1;
    private enum Neighbors { TOP, RIGHT, BOTTOM, LEFT }
    private int n;
    private int totalSiteCount;
    private int[][] sites;
    private WeightedQuickUnionUF connectedSites;
    private int virtualTopSiteId;
    private int virtualBottomSiteId;

    // create N-by-N sites, with all sites blocked
    
    public Percolation(int n) {
        if (n <= 0) throw new IllegalArgumentException();
        this.n = n;
        this.totalSiteCount = n * n;
        this.sites = new int[n][n];
        this.connectedSites = new WeightedQuickUnionUF(this.totalSiteCount + 2);
        this.virtualTopSiteId = this.totalSiteCount;
        this.virtualBottomSiteId = this.totalSiteCount + 1;

        for (int i = 0; i < n; i++) {
            this.connectedSites.union(virtualTopSiteId, i);
            this.connectedSites.union(virtualBottomSiteId,
                                      (this.totalSiteCount - 1 - i));
            for (int j = 0; j < n; j++) {
                this.sites[i][j] = BLOCKED;
            }
        }
    }

    // open site (row i, column j) if it is not open already
    public void open(int i, int j) {
        if (i > this.sites.length || j > this.sites.length) 
            throw new IndexOutOfBoundsException();
        if (i <= 0 || j <= 0) throw new IndexOutOfBoundsException();
        int rowIndex = i - 1;
        int columnIndex = j - 1;
        
        // how does this get you the site id?
        int siteId = (rowIndex * this.n) + columnIndex;

        for (Neighbors neighbor: Neighbors.values()) {
            int neighborRowIndex = rowIndex;
            int neighborColumnIndex = columnIndex;
            if (neighbor == Neighbors.TOP) neighborRowIndex--;
            if (neighbor == Neighbors.RIGHT) neighborColumnIndex++;
            if (neighbor == Neighbors.BOTTOM) neighborRowIndex++;
            if (neighbor == Neighbors.LEFT) neighborColumnIndex--;
            if (neighborRowIndex < 0 || neighborRowIndex >= this.n 
                    || neighborColumnIndex < 0 || neighborColumnIndex >= this.n)
                continue;

            if (this.sites[neighborRowIndex][neighborColumnIndex] == OPEN) {
                int neighborSiteId = 
                    (neighborRowIndex * this.n) + neighborColumnIndex;
                if (!connectedSites.connected(siteId, neighborSiteId)) {
                    connectedSites.union(siteId, neighborSiteId);
                }
                //why add one to the neighboring row index? Not sure this step is necessary
                if (isFull(neighborRowIndex + 1, neighborColumnIndex + 1)) {
                    this.connectedSites.union(this.virtualTopSiteId, siteId);
                }
            }
        }

        this.sites[rowIndex][columnIndex] = OPEN;
    }

    // is site (row i, column j) open?
    public boolean isOpen(int i, int j) {
        if (i > this.sites.length || j > this.sites.length) 
            throw new IndexOutOfBoundsException();
        if (i <= 0 || j <= 0) throw new IndexOutOfBoundsException();

        return this.sites[i - 1][j - 1] == OPEN;
    }

    /**
     * A full site is an open site that can be connected to an open site 
     * in the top row via a chain of neighboring
     * (left, right, up, down) open sites.
     */
    public boolean isFull(int i, int j) {
        if (i > this.sites.length || j > this.sites.length) 
            throw new IndexOutOfBoundsException();
        if (i <= 0 || j <= 0) throw new IndexOutOfBoundsException();
        int rowIndex = i - 1;
        int columnIndex = j - 1;
        int siteId = (rowIndex * this.n) + columnIndex;

        return isOpen(i, j) 
            && this.connectedSites.connected(siteId, virtualTopSiteId);
    }

    /**
     * We say the system percolates if there is a full site 
     * in the bottom row. In other words, a system percolates if
     * we fill all open sites connected to the top row and that 
     * process fills some open site on the bottom row.
     */
    public boolean percolates() {
        if (this.totalSiteCount == 1)
            return isFull(1, 1);
        return this.connectedSites.connected(virtualBottomSiteId, virtualTopSiteId);
    }
}